### PR TITLE
[apiap] Polymorphic owner for Metric

### PR DESCRIPTION
**Pull request by guicassolato**
This is step 1 of migrating metrics from the services to the backend APIs. Next step will be removing `service_id` and using only the new polymorphic owner.

Closes [THREESCALE-3158](https://issues.jboss.org/browse/THREESCALE-3158)

**Special notes for your reviewer**
It contains commits of https://github.com/3scale/porta/pull/1059

### CODE REVIEW

File: db/migrate/20190805135829_backfill_metric_owners.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class BackfillMetricOwners < ActiveRecord::Migration
+  def up
+    say_with_time 'Updating metric owners...' do
+      Metric.reset_column_information
```
Comment (hallelujah) [inline db/migrate/20190805135829_backfill_metric_owners.rb]: Yes as the previous migration will be run at the same time as this one, we need to reset the column information of the model otherwise it will not know about the newly created columns

File: db/schema.rb [hunk 2]
```diff
@@ -1155,8 +1158,11 @@
     t.text     "redirect_url",       limit: 65535
     t.integer  "position",           limit: 4
     t.boolean  "last",                             default: false
+    t.integer  "owner_id",           limit: 8
```
Comment (hallelujah) [inline db/schema.rb]: This does not belong to this PR

File: db/schema.rb [hunk 3]
```diff
@@ -1228,6 +1234,7 @@
     t.boolean  "referrer_filters_required",                    default: false
     t.string   "deployment_option",              limit: 255,   default: "hosted"
     t.string   "kubernetes_service_link",        limit: 255
+    t.boolean  "act_as_product",                               default: false
```
Comment (hallelujah) [inline db/schema.rb]: This does not belong to this PR

File: test/unit/metric_test.rb [hunk 4]
```diff
@@ -72,6 +72,21 @@ def test_destroyable?
...
+    service = FactoryBot.create(:service, account: provider)
+    service_metric = FactoryBot.build(:metric, service: service)
+    refute service_metric.owner
+    assert service_metric.valid?
+    assert_equal service, service_metric.owner
+
+    backend_api = BackendApi.create(name: 'API', system_name: 'api', account: provider)
+    backend_metric = FactoryBot.build(:metric, owner: backend_api)
+    assert_equal backend_api, backend_metric.owner
+    assert service_metric.valid?
```
Comment (guicassolato) [inline test/unit/metric_test.rb]:
```suggestion
    assert backend_metric.valid?
```

File: db/oracle_schema.rb [hunk 5]
```diff
@@ -817,14 +817,17 @@
...
+    t.datetime "created_at",                             null: false
     t.datetime "updated_at"
-    t.integer  "service_id",    precision: 38
+    t.integer  "service_id",              precision: 38
     t.string   "friendly_name"
-    t.integer  "parent_id",     precision: 38
-    t.integer  "tenant_id",     precision: 38
+    t.integer  "parent_id",               precision: 38
+    t.integer  "tenant_id",               precision: 38
+    t.integer  "owner_id",      limit: 8, precision: 8
```
Comment (hallelujah) [inline db/oracle_schema.rb]: As @Martouta noticed, this should be only`precision: 38`

File: test/unit/metric_test.rb [hunk 6]
```diff
@@ -72,6 +72,21 @@ def test_destroyable?
     assert metric_method.errors[:system_name].present?
   end
 
+  test 'fill owner' do
+    provider = FactoryBot.create(:simple_provider)
+
+    service = FactoryBot.create(:service, account: provider)
```
Comment (Martouta) [inline test/unit/metric_test.rb]: The only difference between `:service` and `:simple_service` is that `:service` uses a `:provider_account` and `:simple_service` uses a `:simple_provider`, so a possible improvement would be removing the previous line (`provider = FactoryBot.create(:simple_provider)`), here using `service = FactoryBot.create(:simple_service)` and in line 84 doing `backend_api = BackendApi.create(name: 'API', system_name: 'api', account: service.provider)`

But this is completely optional 😄 And in fact I don't think you should repush if this is the only change bcz it is not important at all and we don't have much time 😄
